src/views/virtualmachines/details/tabs/disk/modal/DeleteDiskModal.tsx

```diff
@@ -0,0 +1,135 @@
+import * as React from 'react';
+import produce from 'immer';
+import { printableVMStatus } from 'src/views/virtualmachines/utils';
+
+import DataVolumeModel from '@kubevirt-ui/kubevirt-api/console/models/DataVolumeModel';
+import VirtualMachineModel from '@kubevirt-ui/kubevirt-api/console/models/VirtualMachineModel';
+import { V1VirtualMachine, V1Volume } from '@kubevirt-ui/kubevirt-api/kubevirt';
+import DeleteResourceMessage from '@kubevirt-utils/components/DeleteResourceMessage/DeleteResourceMessage';
+import { getRemoveHotplugPromise } from '@kubevirt-utils/components/DiskModal/utils/helpers';
+import Loading from '@kubevirt-utils/components/Loading/Loading';
+import TabModal from '@kubevirt-utils/components/TabModal/TabModal';
+import { useKubevirtTranslation } from '@kubevirt-utils/hooks/useKubevirtTranslation';
+import { buildOwnerReference, compareOwnerReferences } from '@kubevirt-utils/resources/shared';
+import { getDataVolumeTemplates, getDisks, getVolumes } from '@kubevirt-utils/resources/vm';
+import { k8sDelete, K8sResourceCommon, k8sUpdate } from '@openshift-console/dynamic-plugin-sdk';
+import { ButtonVariant, Checkbox, Stack, StackItem } from '@patternfly/react-core';
+
+import useVolumeOwnedResource from './hooks/useVolumeOwnedResource';
+
+type DeleteDiskModalProps = {
+  vm: V1VirtualMachine;
+  volume: V1Volume;
+  isOpen: boolean;
+  onClose: () => void;
+};
+
+const DeleteDiskModal: React.FC<DeleteDiskModalProps> = ({ vm, volume, isOpen, onClose }) => {
+  const { t } = useKubevirtTranslation();
+  const [deleteOwnedResource, setDeleteOwnedResource] = React.useState(true);
+
+  const diskName = volume?.name;
+
+  const { volumeResource, loaded, volumeResourceName, volumeResourceModel } =
+    useVolumeOwnedResource(vm, volume);
+
+  const updatedVirtualMachine = React.useMemo(() => {
+    const updatedDisks = getDisks(vm)?.filter(({ name }) => name !== diskName);
+    const updatedVolumes = getVolumes(vm)?.filter(({ name }) => name !== diskName);
+    const updatedDataVolumeTemplates = getDataVolumeTemplates(vm)?.filter(
+      (dvt) => dvt?.metadata?.name !== volumeResourceName,
+    );
+
+    const updatedVM = produce<V1VirtualMachine>(vm, (vmDraft: V1VirtualMachine) => {
+      vmDraft.spec.template.spec.domain.devices.disks = updatedDisks;
+      vmDraft.spec.template.spec.volumes = updatedVolumes;
+      vmDraft.spec.dataVolumeTemplates = updatedDataVolumeTemplates;
+    });
+    return updatedVM;
+  }, [vm, diskName, volumeResourceName]);
+
+  const onSubmit = (updatedVM: K8sResourceCommon) => {
```

Comment:
`(updatedVM: V1VirtualMachine)`

Reply:
if you look closly, eventually the last k8s call I'm making is either deleteing the volume's resource (when checkbox is checked) `volumeResource` which is of type `K8sResourceCommon` OR I update the resource to remove the VM's ownerReference for `updatedResourceVolume` which is also typed as `K8sResourceCommon` 
long story short, changing this will cause a type error with the onSubmit prop on TabModal component